applications/nrf5340_audio/src/audio/sw_codec_select.c

```diff
@@ -49,33 +61,109 @@ int sw_codec_encode(void *pcm_data, size_t pcm_size, uint8_t **encoded_data, siz
 		 * stereo PCM stream
 		 */
 		ret = pscm_two_channel_split(pcm_data, pcm_size, CONFIG_AUDIO_BIT_DEPTH_BITS,
-					     pcm_data_mono[AUDIO_CH_L], pcm_data_mono[AUDIO_CH_R],
-					     &pcm_block_size_mono);
+					     pcm_data_mono_system_sample_rate[AUDIO_CH_L],
+					     pcm_data_mono_system_sample_rate[AUDIO_CH_R],
+					     &pcm_block_size_mono_system_sample_rate);
 		if (ret) {
 			return ret;
 		}
 
 		switch (m_config.encoder.channel_mode) {
 		case SW_CODEC_MONO: {
-			ret = sw_codec_lc3_enc_run(pcm_data_mono[m_config.encoder.audio_ch],
-						   pcm_block_size_mono, LC3_USE_BITRATE_FROM_INIT,
-						   0, sizeof(m_encoded_data), m_encoded_data,
+			char *lc3_in_data_left_ptr;
+
+			if (m_config.encoder.sample_rate_hz == CONFIG_AUDIO_SAMPLE_RATE_HZ) {
+				lc3_in_data_left_ptr = pcm_data_mono_system_sample_rate[AUDIO_CH_L];
+				pcm_block_size_mono = pcm_block_size_mono_system_sample_rate;
+			} else {
+#ifdef CONFIG_AUDIO_SAMPLE_RATE_CONVERSION
+				ret = sample_rate_converter_process(
+					&sample_rate_converter_left, SAMPLE_RATE_FILTER_TEST,
+					(void *)pcm_data_mono_system_sample_rate[AUDIO_CH_L],
+					pcm_block_size_mono_system_sample_rate,
+					CONFIG_AUDIO_SAMPLE_RATE_HZ,
+					(void *)pcm_data_mono[AUDIO_CH_L],
+					ARRAY_SIZE(pcm_data_mono[AUDIO_CH_L]), &pcm_block_size_mono,
+					m_config.encoder.sample_rate_hz);
+				if (ret) {
+					LOG_ERR("Sample rate conversion failed for left channel: "
+						"%d",
+						ret);
+					return ret;
+				}
+				lc3_in_data_left_ptr = pcm_data_mono[AUDIO_CH_L];
+#else
+				LOG_ERR("Sample rates are not equal");
+				return -EINVAL;
+#endif /* CONFIG_AUDIO_SAMPLE_RATE_CONVERSION */
+			}
```

Comment:
This part is similar for mono and stereo, right? Only that stereo does it for both channels. I suggest moving it up and then do it regardless, but guard the sample rate conversion for right channel with a SW_CODEC_STEREO check.  Would remove a lot of duplicated code.

Reply:
Furthermore, you could do it in a for-loop that looks like:
```C
for (int i = 0; i < m_config.encoder.channel_mode; i++) 
```
and then use i as array index for pcm_data_mono etc, since AUDIO_CH_L==0 and AUDIO_CH_R==1.
You would have to change sample_rate_converter_ctx to an array of pointers to the structs and use the same index for them.

It sounds like a lot of changes, but it's really not and I think the code would look much nicer